### PR TITLE
[interpreter] Add ref type and call_ref instruction

### PR DESCRIPTION
**Pull request by rossberg**
As described in Overview:

* Introduce `(ref $t)` type
* Introduce `def_type` for entries in type section (currently only func types)
* Extend type equivalence and matching
* Add `call_ref` and `return_call_ref` instructions
  (with ReturnInvoking administrative instruction from tail-call proposal)
* Disallow non-defaultable local or table types
* Some tests

Various TODOs where types need to be compared outside a module, i.e., with no type section present. We need an internal representation for ref types that does not depend on type indices, e.g., using a type store.

### CODE REVIEW

File: proposals/function-references/Overview.md [hunk 1]
```diff
@@ -124,14 +124,14 @@ Based on [reference types proposal](https://github.com/WebAssembly/reference-typ
...
-  - `(opt? $t) ok` iff `$t` is defined in the context
+* `constype ::= null? <typeidx> | any | func | null`
+  - `(null? $t) ok` iff `$t` is defined in the context
   - `any ok` and `func ok` and `null ok`, always
   - Note: `null` is only on internal type
 
 * In the binary encoding,
-  - the non-optional `<typeidx>` type is encoded as a (positive) signed LEB
-  - `opt` is given a new (negative) type opcode, follosed by the unsigned LEB
+  - the non-nulable `<typeidx>` type is encoded as a (positive) signed LEB
```
Comment (binji) [inline proposals/function-references/Overview.md]: nullable

File: proposals/function-references/Overview.md [hunk 2]
```diff
@@ -124,14 +124,14 @@ Based on [reference types proposal](https://github.com/WebAssembly/reference-typ
...
+  - `(null <typidx>)` given a new (negative) type opcode, follosed by the unsigned LEB
```
Comment (binji) [inline proposals/function-references/Overview.md]: followed

File: test/core/binary.wast [hunk 3]
```diff
@@ -1491,7 +1491,7 @@
     "\02"                                   ;; break depth for default
     "\0b\0b\0b"                             ;; end
   )
-  "invalid reference type"
+  "unexpected end of section or function"
```
Comment (binji) [inline test/core/binary.wast]: Seems like this is broken upstream, it shouldn't be "invalid reference type", right?

File: test/core/call_ref.wast [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    (call_ref (local.get $x) (local.get $f))
+  )
+
+  (func $f (type $ii) (i32.mul (local.get 0) (local.get 0)))
+  (func $g (type $ii) (i32.sub (i32.const 0) (local.get 0)))
+
+  (elem func $f $g)
+
+  (func (export "run") (param $x i32) (result i32)
+    (local $rf (ref null $ii))
```
Comment (binji) [inline test/core/call_ref.wast]: I see why `(ref null ...)` is more consistent, but it looks odd here. It's harder to tell that this is nullable, and not always null.

File: proposals/function-references/Overview.md (hunk 1 above)
Comment (rossberg) [reply proposals/function-references/Overview.md]: Done.

File: proposals/function-references/Overview.md (hunk 2 above)
Comment (rossberg) [reply proposals/function-references/Overview.md]: Done.

File: test/core/binary.wast (hunk 3 above)
Comment (rossberg) [reply test/core/binary.wast]: "Invalid value type" would be more appropriate, but this is just coincidence: the surplus label index `\02` happens to decode into a block instruction, that expects a block type next, and `\0b` isn't a numeric type, so the interpreter complaints that it was an invalid reference type. With this PR the same `\0b` is now interpreted as the (syntactically fine) reference type with index 11, and a later error occurs instead.

This is a frequent problem with negative tests in binary.wast: when you add some new opcodes error messages can change somewhat arbitrarily.

File: test/core/call_ref.wast (hunk 4 above)
Comment (rossberg) [reply test/core/call_ref.wast]: When you read it as a union type, i.e., ref(null ∪ $t), it actually makes sense. :)

**Pull request merged**